[Crash]  TabTrayFragment.java line 195

**Describe the bug**
crashes 636 ; user affected 129

https://console.firebase.google.com/u/0/project/zerda-dcf76/crashlytics/app/android:org.mozilla.rocket/issues/844c4fc80faf3a0244a9d171894a7a23?time=last-seven-days&sessionId=5EA981D3019700013C95C76274423FEA_DNE_0_v2


**Device info (please complete the following information):**
![螢幕快照 2020-04-29 下午9 56 18](https://user-images.githubusercontent.com/52307187/80604434-4e1b5c00-8a64-11ea-8e35-5a52c523b874.png)
![螢幕快照 2020-04-29 下午9 56 24](https://user-images.githubusercontent.com/52307187/80604439-4fe51f80-8a64-11ea-90c6-5a19d7825197.png)

This only happens on Version 2.1.16(19398)

I've reproduced this issue twice
1. Launch app without [Don't keep Activity]
2. Open to Tab Tray
3. Go to [System Settings] and enable [Don't keep Activity]
4. Go back to FxLite in Recent Apps
5. App crashes


```
2020-04-30 01:12:27.517 11356-11356/org.mozilla.rocket.debug.nechen E/AndroidRuntime: FATAL EXCEPTION: main
    Process: org.mozilla.rocket.debug.nechen, PID: 11356
    java.lang.RuntimeException: Unable to start activity ComponentInfo{org.mozilla.rocket.debug.nechen/org.mozilla.focus.activity.MainActivity}: java.lang.RuntimeException: Cannot create an instance of class org.mozilla.rocket.home.HomeViewModel
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2817)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2892)
        at android.app.ActivityThread.-wrap11(Unknown Source:0)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1593)
        at android.os.Handler.dispatchMessage(Handler.java:105)
        at android.os.Looper.loop(Looper.java:164)
        at android.app.ActivityThread.main(ActivityThread.java:6541)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.Zygote$MethodAndArgsCaller.run(Zygote.java:240)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:767)
     Caused by: java.lang.RuntimeException: Cannot create an instance of class org.mozilla.rocket.home.HomeViewModel
        at androidx.lifecycle.ViewModelProvider$NewInstanceFactory.create(ViewModelProvider.java:221)
        at androidx.lifecycle.ViewModelProvider.get(ViewModelProvider.java:187)
        at androidx.lifecycle.ViewModelProvider.get(ViewModelProvider.java:150)
        at org.mozilla.focus.tabs.tabtray.TabTrayFragment.onCreateView(TabTrayFragment.java:195)
        at androidx.fragment.app.Fragment.performCreateView(Fragment.java:2698)
        at androidx.fragment.app.FragmentStateManager.createView(FragmentStateManager.java:320)
        at androidx.fragment.app.FragmentManager.moveToState(FragmentManager.java:1187)
        at androidx.fragment.app.FragmentManager.moveToState(FragmentManager.java:1356)
        at androidx.fragment.app.FragmentManager.moveFragmentToExpectedState(FragmentManager.java:1434)
        at androidx.fragment.app.FragmentManager.moveToState(FragmentManager.java:1497)
        at androidx.fragment.app.FragmentManager.dispatchStateChange(FragmentManager.java:2625)
        at androidx.fragment.app.FragmentManager.dispatchActivityCreated(FragmentManager.java:2577)
        at androidx.fragment.app.FragmentController.dispatchActivityCreated(FragmentController.java:247)
        at androidx.fragment.app.FragmentActivity.onStart(FragmentActivity.java:541)
        at androidx.appcompat.app.AppCompatActivity.onStart(AppCompatActivity.java:201)
        at org.mozilla.focus.activity.MainActivity.onStart(MainActivity.kt:344)
        at android.app.Instrumentation.callActivityOnStart(Instrumentation.java:1333)
        at android.app.Activity.performStart(Activity.java:6992)
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2780)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2892) 
        at android.app.ActivityThread.-wrap11(Unknown Source:0) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1593) 
        at android.os.Handler.dispatchMessage(Handler.java:105) 
        at android.os.Looper.loop(Looper.java:164) 
        at android.app.ActivityThread.main(ActivityThread.java:6541) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.Zygote$MethodAndArgsCaller.run(Zygote.java:240) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:767) 
     Caused by: java.lang.InstantiationException: java.lang.Class<org.mozilla.rocket.home.HomeViewModel> has no zero argument constructor
        at java.lang.Class.newInstance(Native Method)
        at androidx.lifecycle.ViewModelProvider$NewInstanceFactory.create(ViewModelProvider.java:219)
        at androidx.lifecycle.ViewModelProvider.get(ViewModelProvider.java:187) 
        at androidx.lifecycle.ViewModelProvider.get(ViewModelProvider.java:150) 
        at org.mozilla.focus.tabs.tabtray.TabTrayFragment.onCreateView(TabTrayFragment.java:195) 
        at androidx.fragment.app.Fragment.performCreateView(Fragment.java:2698) 
        at androidx.fragment.app.FragmentStateManager.createView(FragmentStateManager.java:320) 
        at androidx.fragment.app.FragmentManager.moveToState(FragmentManager.java:1187) 
        at androidx.fragment.app.FragmentManager.moveToState(FragmentManager.java:1356) 
        at androidx.fragment.app.FragmentManager.moveFragmentToExpectedState(FragmentManager.java:1434) 
        at androidx.fragment.app.FragmentManager.moveToState(FragmentManager.java:1497) 
        at androidx.fragment.app.FragmentManager.dispatchStateChange(FragmentManager.java:2625) 
        at androidx.fragment.app.FragmentManager.dispatchActivityCreated(FragmentManager.java:2577) 
        at androidx.fragment.app.FragmentController.dispatchActivityCreated(FragmentController.java:247) 
        at androidx.fragment.app.FragmentActivity.onStart(FragmentActivity.java:541) 
        at androidx.appcompat.app.AppCompatActivity.onStart(AppCompatActivity.java:201) 
        at org.mozilla.focus.activity.MainActivity.onStart(MainActivity.kt:344) 
        at android.app.Instrumentation.callActivityOnStart(Instrumentation.java:1333) 
        at android.app.Activity.performStart(Activity.java:6992) 
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2780) 
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2892) 
        at android.app.ActivityThread.-wrap11(Unknown Source:0) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1593) 
        at android.os.Handler.dispatchMessage(Handler.java:105) 
        at android.os.Looper.loop(Looper.java:164) 
        at android.app.ActivityThread.main(ActivityThread.java:6541) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.Zygote$MethodAndArgsCaller.run(Zygote.java:240) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:767) 
```

--- a/app/src/main/java/org/mozilla/focus/tabs/tabtray/TabTrayFragment.java
+++ b/app/src/main/java/org/mozilla/focus/tabs/tabtray/TabTrayFragment.java
@@ -54,6 +54,8 @@ import org.mozilla.focus.navigation.ScreenNavigator;
 import org.mozilla.focus.telemetry.TelemetryWrapper;
 import org.mozilla.focus.utils.Settings;
 import org.mozilla.focus.utils.ViewUtils;
+import org.mozilla.rocket.content.BaseViewModelFactory;
+import org.mozilla.rocket.content.ExtentionKt;
 import org.mozilla.rocket.home.HomeViewModel;
 import org.mozilla.rocket.nightmode.themed.ThemedImageView;
 import org.mozilla.rocket.nightmode.themed.ThemedRecyclerView;
@@ -67,6 +69,10 @@ import org.mozilla.rocket.tabs.SessionManager;
 import org.mozilla.rocket.tabs.TabsSessionProvider;
 
 import java.util.List;
+
+import javax.inject.Inject;
+
+import dagger.Lazy;
 
 public class TabTrayFragment extends DialogFragment implements TabTrayContract.View,
         View.OnClickListener, TabTrayAdapter.TabClickListener {
@@ -107,6 +113,8 @@ public class TabTrayFragment extends DialogFragment implements TabTrayContract.V
     private Runnable dismissRunnable = this::dismissAllowingStateLoss;
 
     private TabTrayViewModel tabTrayViewModel = null;
+    @Inject
+    Lazy<HomeViewModel> homeViewModelCreator;
     private HomeViewModel homeViewModel = null;
 
     private ThemedImageView imgPrivateBrowsing, imgNewTab;
@@ -124,6 +132,7 @@ public class TabTrayFragment extends DialogFragment implements TabTrayContract.V
 
     @Override
     public void onCreate(@Nullable Bundle savedInstanceState) {
+        ExtentionKt.appComponent(this).inject(this);
         super.onCreate(savedInstanceState);
         setStyle(DialogFragment.STYLE_NO_TITLE, R.style.TabTrayTheme);
 
@@ -192,7 +201,7 @@ public class TabTrayFragment extends DialogFragment implements TabTrayContract.V
                 adapter.setShoppingSearch(showShoppingSearch, uiModel.getKeyword());
             }
         });
-        homeViewModel = new ViewModelProvider(getActivity(), new ViewModelProvider.NewInstanceFactory()).get(HomeViewModel.class);
+        homeViewModel = new ViewModelProvider(requireActivity(), new BaseViewModelFactory<>(homeViewModelCreator::get)).get(HomeViewModel.class);
         backgroundView = view.findViewById(R.id.root_layout);
         logoMan = backgroundView.findViewById(R.id.logo_man);
         imgPrivateBrowsing = view.findViewById(R.id.img_private_browsing);

--- a/app/src/main/java/org/mozilla/rocket/di/AppComponent.kt
+++ b/app/src/main/java/org/mozilla/rocket/di/AppComponent.kt
@@ -27,6 +27,7 @@ import org.mozilla.focus.fragment.DownloadsFragment
 import org.mozilla.focus.history.BrowsingHistoryFragment
 import org.mozilla.focus.persistence.TabsDatabase
 import org.mozilla.focus.persistence.di.TabsModule
+import org.mozilla.focus.tabs.tabtray.TabTrayFragment
 import org.mozilla.focus.urlinput.UrlInputFragment
 import org.mozilla.rocket.chrome.ChromeViewModel
 import org.mozilla.rocket.chrome.di.ChromeModule
@@ -107,6 +108,7 @@ interface AppComponent {
     fun inject(browserFragment: org.mozilla.rocket.privately.browse.BrowserFragment)
     fun inject(downloadsFragment: DownloadsFragment)
     fun inject(homeFragment: HomeFragment)
+    fun inject(tabTrayFragment: TabTrayFragment)
     fun inject(privateHomeFragment: PrivateHomeFragment)
     fun inject(urlInputFragment: UrlInputFragment)
     fun inject(menuDialog: MenuDialog)

